.github/workflows/pull_upstream.yml

```diff
@@ -0,0 +1,82 @@
+name: Pull from Upstream
+
+on:
+  workflow_dispatch:
+    inputs:
+      upstream_repo:
+        description: 'Select Upstream Repository'
+        required: true
+        type: choice
+        options:
+          - 'openvinotoolkit/model_server'
+          - 'kserve/kserve'
+          - 'kserve/modelmesh'
+          - 'caikit-tgis-serving'
+          - 'openvinotoolkit/openvino'
+          - 'vllm-project/vllm'
+          - 'caikit/caikit-nlp'
+          - 'caikit/caikit'
+          - 'odh-model-controller'
+          - 'caikit/caikit-tgis-backend'
+          - 'caikit-nlp-client'
+          - 'kubeflow/model-registry'
+      sync_repo:
+        description: 'Select Repository to sync'
+        required: true
+        type: choice
+        options:
+          - 'Upstream->Midstream(opendatahub-io)'
+          - 'Midstream(opendatahub-io)->Downstream(red-hat-data-services)'
+      source_branch:
+        description: 'Source branch to pull from'
+        required: true
+      target_branch:
+        description: 'Target branch to pull into'
+        required: true
+      force_sync:
+        description: 'Force sync the repository'
+        required: false
+        type: choice
+        options:
+          - 'true'
+          - 'false'
+
+jobs:
+  pull:
+    runs-on: ubuntu-latest
+    outputs: 
+        target_repo: ${{ steps.set-repo.outputs.target_org_repo }}
+    steps:
+    - name: Set repository
+      id: set-repo
+      run: |
+        if [[ "${{ github.event.inputs.upstream_repo }}" == *"/"* ]]; then
+          IFS='/' read -r upstreamOwner repoName <<< "${{ github.event.inputs.upstream_repo }}"
+        else
+          repoName=${{ github.event.inputs.upstream_repo }}
+          upstreamOwner=""
+        fi
+        if [[ "$repoName" == "model_server" ]]; then
+          repoName=""openvino_"$repoName"
+        fi
+        if [ -n "$upstreamOwner" ] && [[ "${{ github.event.inputs.sync_repo }}" == "Upstream->Midstream(opendatahub-io)" ]]; then
+          echo "upstream_org_repo=${{ github.event.inputs.upstream_repo }}"  >> $GITHUB_OUTPUT
+          echo "target_org_repo=opendatahub-io/$repoName" >> $GITHUB_OUTPUT 
+        else
+          echo "upstream_org_repo=opendatahub-io/$repoName"  >> $GITHUB_OUTPUT
+          echo "target_org_repo=red-hat-data-services/$repoName" >> $GITHUB_OUTPUT
+        fi
+    - name: Set git config
+      run: |
+          git config --global user.name "github-actions"
+          git config --global user.email "github-actions@github.com"
+    - name: sync
+      uses: TobKed/github-forks-sync-action@master

```

Comment:
I realize that I'm a bit late to this, but looking at the source for this workflow:

```bash
git clone "${upstream_repo}" -b "${INPUT_UPSTREAM_BRANCH}" --single-branch
cd "${upstream_dir}"
# shellcheck disable=SC2086
git push $_FORCE_OPTION $_TAGS "${target_repo}" "${INPUT_UPSTREAM_BRANCH}:${INPUT_TARGET_BRANCH}"
```

This should work fine if the target repo does not have any additional commits compared to upstream, but it will fail otherwise:

```
 ! [rejected]        main -> main (fetch first)
error: failed to push some refs to '<fork>'
hint: Updates were rejected because the remote contains work that you do not
hint: have locally. This is usually caused by another repository pushing to
hint: the same ref. If you want to integrate the remote changes, use
hint: 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

For this we need to:
1. Clone the fork repo
2. Check out the fork branch that needs to be synced
2. Add the upstream repo as a remote and fetch from it
3 Create a merge commit. `git merge --no-ff upstream/<target branch>` 
4. Either create a pull request or push to the fork.

The current strategy I use for [odh/vllm][https://github.com/opendatahub-io/vllm] is the following, which uses the github cli to directly create a PR from the upstream target branch to the fork target branch. 

https://github.com/opendatahub-io/vllm/blob/main/.github/workflows/sync-with-upstream.yml

Note that this works only when there's no conflicts: in that case wee need to manually solve the conflicts and perform the steps listed above.

Reply:
@dtrifiro will look into that